frontend/src/metabase/parameters/components/widgets/NumberInputWidget/NumberInputWidget.tsx

```diff
@@ -58,16 +59,63 @@ export function NumberInputWidget({
     }
   };
 
-  function shouldCreate(value: string | number) {
-    const res = parseNumberValue(value);
-    return res !== null && res.toString() === value;
-  }
-
   const filteredUnsavedArrayValue = useMemo(
     () => unsavedArrayValue.filter((x): x is number => x !== undefined),
     [unsavedArrayValue],
   );
 
+  const values = parameter?.values_source_config?.values ?? [];
+  const options =
+    values.map(getOption).filter((item): item is SelectItem => item !== null) ??
+    [];
+
+  const valueOptions = unsavedArrayValue
+    .map((item): SelectItem | null => {
+      const option = parameter?.values_source_config?.values?.find(
+        option => getValue(option)?.toString() === item?.toString(),
+      );
+
+      if (!option) {
+        return null;
+      }
+
+      const value = getValue(option)?.toString();
+      if (typeof value !== "string") {
+        return null;
+      }
+
+      return {
+        label: getLabel(option),
+        value,
+      };
+    })
+    .filter(isNotNull);
+
+  const customLabelOptions = options.filter(
+    option => option.label !== option.value,
+  );
+
+  function parseValue(value: string | number | undefined): number | null {
+    if (value === undefined) {
+      return null;
+    }
+
+    const opt = options.find(option => {
```

Comment:
We shouldn't do this. We should not accept free-form input from remapped values. You need to select a value from the list. Doing it otherwise makes it inconsistent with all other places with remapped values.

Reply:
I've removed the match.